vars/googleCloudFunctionDeploy.groovy

```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/groovy
+
+/**
+ * Deploy a Google Cloud Function (HTTP or PubSub triggered)
+ * When using this remember to configure discarding artifacts so we don't hoard them
+ *
+ * CLI docs: https://cloud.google.com/sdk/gcloud/reference/functions/deploy
+ *
+ * @param config A map containing configuration parameters for the function deployment:
+ *   functionName        : Name of the Cloud Function (String, required)
+ *   projectId           : Google Cloud Project ID (String, required)
+ *                         - make sure the project has Jenkins worker as a principal (currently for qa and prod)
+ *   gcsPath             : Google Cloud Storage path for the function archive (String, required)
+ *   runtime             : Runtime for the Cloud Function (String, required)
+ *   region              : Deployment region (String, required)
+ *   serviceAccountEmail : Service account email for the function (String, required)
+ *   environmentVariables: Map of environment variables for the function (Map, required)
+ *   entryPoint          : Entry point for the function (String, optional)
+ *   timeout             : Function timeout (String, optional)
+ *   maxInstances        : Maximum number of function instances (Integer, optional)
+ *   triggerType         : Type of trigger - 'http' or 'pubsub' (String, required)
+ *   topicName           : PubSub topic name (required if triggerType is 'pubsub')
+ *   generation          : Cloud Function generation - 'gen1' or 'gen2' (String, required)
```

Comment:
Gen2 is somewhat more expensive since it requires a few more apis to be enabled (https://www.reddit.com/r/googlecloud/comments/17qw7p6/whats_the_point_of_cloud_function_2nd_gen/, https://cloud.google.com/functions/pricing). For most of our functions Gen1 is more than enough and free (unless we go over the free tier invocation limit).

That is why, in my opinion, using 'gen2' should be harder and very intentional to whomever is deploying functions.

For context we currently have only 1 gen2 function running, which Tshepo uses since it can do concurrent requests. If the gen1:gen2 ratio becomes <1 we could change this.